nim/app.py

```diff
@@ -34,12 +36,15 @@ def update_game_state():
 
 @app.route('/new', methods=['GET', 'POST'])
 def new_game():
-    json = loads('{"min": 3, "max": 12, "piles": 3}')
-    state = app.game.new_game(
-        min=json['min'],
-        max=json['max'],
-        piles=json['piles'],
-    )
+    game_request = request.get_json()
+    try:
+        state = app.game.new_game(
+            min=game_request['min'],
```

Comment:
We can get a key error here the the JSON is incomplete.
`min=game_request.get('min') or session.get('min') or DEFAULT_MIN`
Same for `'max'` and `'piles'`

Reply:
I think part of that logic should be in `new_game()`. I don't want to always pass `min`, `max` or `piles`, so they should take a default value.
Once that's done, here we could do:
```
    try:
        params = {param: game_request[param]
                  for param in ("min", "max", "piles")
                  if param in game_request}
        state = app.game.new_game(**params)
    except NimException as e:
        return jsonify({'error': e.args})
```
I'm not sure if we should support passing params through the `session`. Why do you want to add that?

Reply:
If we refresh the page everything will go back to default values. If we store their preferences in the session, then people can refresh the page, close and reopen, &c without having to reset all of their preferences every time. This is some super necessary behavior for the front end.

At this point we are only passing game preferences so that comprehension isn't doing anything.
```
In [1]: foo = {'a': 1, 'b': 2}

In [2]: bar = {param: foo[param] for param in ('a', 'b', 'c') if param in foo}

In [3]: bar
Out[3]: {'a': 1, 'b': 2}

In [4]: bar == foo
Out[4]: True
```
 I do like your idea of delegating logic to `new_game`. Since when we unpack dictionaries, the second overwrites the first, we can do...
```
params = {**session.get('game_prefs', {}), **game_request}
state = app.game.new_game(**params)
```
And if `game_request` ever becomes more complicated...
```
params = {**session.get('game_prefs', {}), **game_request.get('game_prefs', {})}
state = app.game.new_game(**params)
```

Reply:
Holy fuck that's beautiful! So your brain is functioning at 100% even at 5am huh?

Reply:
Hahahahahahaha :)

Reply:
So I first looked at @dodecaphonix code yesterday. I knew then it was beautiful.. but I didn't know how much I didn't know.
I was almost entirely unfamiliar with unpacking using * or ** and what you could do with it, especially in the case of dictionaries. After looking up examples, wonderful conversations & help from the man himself (lookin' at you @dodecaphonix ) & playing around with them a bit today, I am starting to grasp just how wonderful this code is. 
I've started implementing this into the new_game() route in Nim and so far it's working beautifully. 
Thanks to both of you for the feedback! 
More to follow 🙂

Reply:
@dodecaphonix I do have one question regarding `params = {**session.get('game_prefs', {}), **game_request}` 
Why are we setting an empty dict if there is no `game_prefs` ? 
Is there a difference between game_prefs and the defaults we want to take the place of any field not passed? 
Because if they are the same, game_prefs will be a static line set in the code, which would never be empty. We could then update it to whatever when the user sends the arguments. 
Unless we think the user could put this in a bad state, why would we need to be afraid of this ever being blank? 

Alternatively, if game_prefs *is* a different thing to defaults set if there is an unnamed parameters arguments, I see why you're doing this.